### PR TITLE
[MLA-1880] Raycast sensor interface improvements

### PR DESCRIPTION
**Pull request by chriselion**
### Proposed change(s)
Public-facing change: 
* Adds `RayPerceptionSensor.RayPerceptionOutput`, which contains the most recent ray hit information. This can be used to for additional game logic as needed.
* Adds new fields to `RayOutput`: `StartPositionWorld`, `EndPositionWorld`, `ScaledRayLength`, and `ScaledCastRadius`.

Internal changes:
* Get rid of DebugDisplayInfo - all of its info is now on `RayOutput`, except for the frame that the results were queried, which is now stored as `m_DebugLastFrameCount`
* Raycasts now happen in the Update() call, which always happens before the Write().

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)
Based on feedback from several forum posts:

Forum threads:

- ["i want to do an if statement inside my agent script to say if the ray cast hits an object with the tag enemy then do something"](https://forum.unity.com/threads/getting-the-detectable-tags-from-ray-perception-sensor-script.834733/)
- ["does the ray perception 3d component get the distance to the object it impacted? I need my agents to get the distance to that object"](https://forum.unity.com/threads/ray-perception-sensor-component-3d-distance.867223/)
- ["we have to reset agent if it follows wrong raycast and crash, or we want to add reward if it follows correct raycast."](https://forum.unity.com/threads/how-can-i-use-rayperceptioncomponent3d-effectively.926426/#post-6086127)
- ["I am trying to teach an agent to find a target but once it has "seen" the target, I want to give it a reward."](https://forum.unity.com/threads/rewarding-an-agent-when-it-sees-something.955029/)
- ["I was wondering, if there is an easy way to draw the Debug gizmos of the Ray Perception Sensor ingame for demonstration purposes?"](https://forum.unity.com/threads/way-to-visualize-ray-perception-sensor.978642/)
- ["I'd like to get a value of distance between the agent and obstacles with using Ray Perception Sensor."](https://forum.unity.com/threads/how-to-get-value-of-distance-between-agent-obstacles-with-using-rayperceptionsensor3d.1043338/)
- ["I use three Ray Perception Sensor Components for measuring distances from the agent to some objects, and I'd like to get all values (HasHit, HitTaggedObject, HitTagIndex and HitFraction) from all components."](https://forum.unity.com/threads/how-to-get-values-from-multiple-ray-perception-sensor-3ds.1045960/)

### Types of change(s)
- [x] New feature
- [x] Code refactor

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)

### CODE REVIEW

File: Project/ProjectSettings/ProjectSettings.asset [hunk 1]
```diff
@@ -125,7 +125,8 @@ PlayerSettings:
     16:9: 1
     Others: 1
   bundleVersion: 1.0
-  preloadedAssets: []
+  preloadedAssets:
+  - {fileID: 0}
```
Comment (dongruoping) [inline Project/ProjectSettings/ProjectSettings.asset:129]: relevant?

File: com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensorComponentBase.cs [hunk 2]
```diff
@@ -252,16 +252,26 @@ internal void UpdateSensor()
...
+                return Time.frameCount - m_RaySensor.DebugLastFrameCount;
+            }
+
+            return 0;
+        }
+
         void OnDrawGizmosSelected()
         {
-            if (m_RaySensor?.debugDisplayInfo?.rayInfos != null)
+            if (m_RaySensor?.m_rayPerceptionOutput?.RayOutputs != null)
```
Comment (dongruoping) [inline com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensorComponentBase.cs:267]:
```suggestion
            if (m_RaySensor?.RayPerceptionOutput?.RayOutputs != null)
```

File: com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensor.cs [hunk 3]
```diff
@@ -241,12 +243,16 @@ public class RayPerceptionSensor : ISensor, IBuiltInSensor
         string m_Name;
 
         RayPerceptionInput m_RayPerceptionInput;
+        internal RayPerceptionOutput m_rayPerceptionOutput;
```
Comment (dongruoping) [inline com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensor.cs:246]:
```suggestion
        internal RayPerceptionOutput m_RayPerceptionOutput;
```

**Review by dongruoping — APPROVED** [review]
Looks good to me. Verified that the gizmos work fine in editor.

File: com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensorComponentBase.cs (hunk 2 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensorComponentBase.cs:267]: Thanks, I'll make `m_{r,R}ayPerceptionOutput` private (and fix the capitalization)

File: Project/ProjectSettings/ProjectSettings.asset (hunk 1 above)
Comment (chriselion) [reply Project/ProjectSettings/ProjectSettings.asset:129]: `git add -u` strikes again.

**Pull request merged**